### ISSUE TITLE
[API] get_all_craft_recipes("") throws error

### ISSUE THREAD

**Issue by asanetargoss** [minetest/minetest#2709]
**minetest.get_all_craft_recipes** throws error when called with an empty string. **minetest.get_craft_recipe** does not suffer this problem.

Error from debug.txt:

```
18:38:09: ERROR[main]: ========== ERROR FROM LUA ===========
18:38:09: ERROR[main]: Failed to load and run script from 
18:38:09: ERROR[main]: D:\Portable Programs\minetest-0.4.12\bin\..\mods\testmod\init.lua:
18:38:09: ERROR[main]: caught (...)
18:38:09: ERROR[main]: stack traceback:
18:38:09: ERROR[main]: 	[C]: in function 'get_all_craft_recipes'
18:38:09: ERROR[main]: 	...ams\minetest-0.4.12\bin\..\mods\testmod\init.lua:2: in main chunk
18:38:09: ERROR[main]: ======= END OF ERROR FROM LUA ========
```

The error occurrs when running this code in init.lua of a mod, with "default" listed as a dependency in depends.txt:

```lua
for registered_item, _ in pairs(minetest.registered_items) do
    local recipes = minetest.get_all_craft_recipes(registered_item)
end
```

I believe the code throws an error because default registers the empty string as an item. I am not sure if that is intended behavior.

Workaround:

```lua
for registered_item, _ in pairs(minetest.registered_items) do
    if registered_item ~= "" then
        local recipes = minetest.get_all_craft_recipes(registered_item)
    end
end
```

**Comment by est31** [comment]
Now the question to begin with, is why there is an item with "" in `minetest.registered_items`?

```
print("Testing 2709")
minetest.get_all_craft_recipes("")
for registered_item, _ in pairs(minetest.registered_items) do
	if registered_item == "" then
		dump("'' found!!!!!!!")
	end
	local recipes = minetest.get_all_craft_recipes(registered_item)
end
```

This doesn't print any error or anything, except the "Testing 2709" line. I've tested with both recent git version and 0.4.12 version of the engine, together with minimal test game and a git version of minetest_game.


The second question is why it fails when you call the method with "", because it doesn't fail for me. 
Can you disable all other mods, and only test with the game? Which game do you have installed?

**Comment by PilzAdam** [comment]
@est31 the item with the empty itemstring is the hand.

**Comment by est31** [comment]
Ah, its been an error in my code, I wrote `dump` instead of `print`. Still I can't reproduce the error @asanetargoss.

**Comment by asanetargoss** [comment]
@est31 The error only happens with "default" listed as a dependency in depends.txt. If I remove the dependency, nothing bad happens. I've tested this with both minetest_game and minimal, both with Minetest 0.4.12. I also verified today that I do get the same behavior when I call the exact same code that you used (stack trace is also the same as before). The only mod loaded was the test mod which I used to run the test code.

I tried a new install and the behavior was still the same. I used the 64-bit MSVC build.

Then I tried a 32-bit MinGW build... and... not even a debug message. Just "Minetest core application has stopped working." Likewise, this behavior only occurs when "default" is listed as a dependency.

I'm using Windows 7, if that helps.

**Comment by est31** [comment]
I can reproduce now.

**Comment by est31** [comment]
It is fixed in master, most likely by commit 03e0dd33a847a83d975282c6caf6b926306e7b57, closing.

**Comment by Zeno-** [comment]
@est31 FYI

```
void func(std::string s)
{
	char *p = NULL;
	if (p == s)
		std::cout << "ok" << std::endl;
}
```
